### PR TITLE
uefi: Append the header on capsules without headers from Linux

### PR DESCRIPTION
**Pull request by superm1**
This allows using better heuristics and potentially phasing this out in the future.

This idea was mentioned in #889 .

It's not yet tested. @hughsie can you arrange for one of the affected machines that need header added to be tested against it?

### CODE REVIEW

**Comment by superm1** [comment]
Alright so I've done some testing on that with the BIOS implementation on an XPS 9370.
* I made sure that it works on capsules that have headers as delivered from LVFS (using FW 1.6.3 that's at LVFS).
* I stripped the header from that capsule and then let fwupd build a new header and verified the BIOS accepts it.

The last commit (b11cc2f) I expect should fix some "small core" systems that have certain page alignment rules.

File: plugins/uefi/fu-uefi-bootmgr.c [hunk 1]
```diff
@@ -108,7 +108,7 @@ fu_uefi_setup_bootnext_with_dp (const guint8 *dp_buf, guint8 *opt, gssize opt_si
 		efidp found_dp;
 		g_autofree guint8 *var_data_tmp = NULL;
 
-		if (efi_guid_cmp (guid, &efi_guid_global))
+		if (efi_guid_cmp (guid, &efi_guid_global) == 0)
```
Comment (hughsie) [inline plugins/uefi/fu-uefi-bootmgr.c]: Although I agree it's correct, what's the ramifications of this issue? i.e. what bugs would it fix?

File: plugins/uefi/fu-uefi-device.c [hunk 2]
```diff
@@ -316,18 +316,19 @@ fu_uefi_device_fixup_firmware (FuDevice *device, GBytes *fw, GError **error)
...
+		guint page_size = getpagesize();
+		guint8 *new_data = g_malloc (fw_length + page_size);
+		guint8 *capsule = new_data + page_size;
 		efi_capsule_header_t *header = (efi_capsule_header_t *) new_data;
 
 		g_warning ("missing or invalid embedded capsule header");
 		header->flags = self->capsule_flags;
-		header->header_size = sizeof(efi_capsule_header_t);
-		header->capsule_image_size = fw_length;
+		header->header_size = page_size;
```
Comment (hughsie) [inline plugins/uefi/fu-uefi-device.c]: is `page_size` always >= `sizeof(efi_capsule_header_t)`?

File: plugins/uefi/fu-uefi-device.c [hunk 3]
```diff
@@ -316,18 +316,19 @@ fu_uefi_device_fixup_firmware (FuDevice *device, GBytes *fw, GError **error)
...
+		header->capsule_image_size = fw_length + page_size;
 		memcpy (&(header->guid), &esrt_guid, sizeof (efi_guid_t));
```
Comment (hughsie) [inline plugins/uefi/fu-uefi-device.c]: no need for the brackets for the first arg to memcpy

File: plugins/uefi/fu-uefi-device.c [hunk 4]
```diff
@@ -289,6 +289,50 @@ fu_uefi_device_build_dp_buf (const gchar *path, gsize *bufsz, GError **error)
...
+	memcpy (&payload_guid, data, sizeof(efi_guid_t));
+
+	/* ESRT header matches payload */
+	if (efi_guid_cmp (&esrt_guid, &payload_guid) == 0) {
+		g_debug ("ESRT matches payload GUID");
+		fixed_fw = g_bytes_new_from_bytes (fw, 0, fw_length);
+	/* FMP payload */
+	} else if (fu_uefi_device_get_kind (self) == FU_UEFI_DEVICE_KIND_FMP) {
+		g_debug ("performing FMP update");
+		fixed_fw = g_bytes_new_from_bytes (fw, 0, fw_length);
```
Comment (hughsie) [inline plugins/uefi/fu-uefi-device.c]: can we just return with the `g_bytes_new_from_bytes` here?

File: plugins/uefi/fu-uefi-device.c [hunk 5]
```diff
@@ -289,6 +289,50 @@ fu_uefi_device_build_dp_buf (const gchar *path, gsize *bufsz, GError **error)
...
+	/* Missing, add a header */
+	} else {
+		guint8 *new_data = g_malloc (fw_length + sizeof(efi_capsule_header_t));
+		guint8 *capsule = new_data + sizeof (efi_capsule_header_t);
+		efi_capsule_header_t *header = (efi_capsule_header_t *) new_data;
+
+		g_warning ("missing or invalid embedded capsule header");
```
Comment (hughsie) [inline plugins/uefi/fu-uefi-device.c]: didn't we want to put this in the uploaded firmware success/failure report somehow?

File: plugins/uefi/fu-uefi-bootmgr.c (hunk 1 above)
Comment (superm1) [reply plugins/uefi/fu-uefi-bootmgr.c]: I think it should speed up iterating entries

File: plugins/uefi/fu-uefi-device.c (hunk 2 above)
Comment (superm1) [reply plugins/uefi/fu-uefi-device.c]: Yes, all architectures have a bigger page than the 27 bytes that `efi_capsule_header_t` is.

File: plugins/uefi/fu-uefi-device.c (hunk 5 above)
Comment (superm1) [reply plugins/uefi/fu-uefi-device.c]: Yes, forgot about that.  Added it.

File: plugins/uefi/fu-uefi-device.c (hunk 3 above)
Comment (superm1) [reply plugins/uefi/fu-uefi-device.c]: Thanks, fixed

File: plugins/uefi/fu-uefi-device.c (hunk 4 above)
Comment (superm1) [reply plugins/uefi/fu-uefi-device.c]: Yep, adjusted.

File: plugins/uefi/fu-uefi-device.c [hunk 6]
```diff
@@ -289,6 +289,50 @@ fu_uefi_device_build_dp_buf (const gchar *path, gsize *bufsz, GError **error)
...
+	efi_guid_t esrt_guid;
+	efi_guid_t payload_guid;
+	const gchar *data = g_bytes_get_data (fw, &fw_length);
+
+	/* convert to EFI GUIDs */
+	if (efi_str_to_guid (fu_uefi_device_get_guid (self), &esrt_guid) < 0) {
+		g_set_error_literal (error, FWUPD_ERROR, FWUPD_ERROR_INTERNAL, "Invalid ESRT GUID");
+		return NULL;
+	}
+	memcpy (&payload_guid, data, sizeof(efi_guid_t));
```
[the remark this answers is not in the document]
Comment (superm1) [reply plugins/uefi/fu-uefi-device.c]: Yes, good catch.

**Comment by hughsie** [comment]
Assuming CI now passes, LGTM, thanks. I'll do some Lenovo testing when one of my machines has an update pending (no downgrades :/)

**Comment by superm1** [comment]
Can you test on Lenovo by just "re-installing" the same payload?

**Comment by hughsie** [comment]
> missing or invalid embedded capsule header

This means we fixed up the file automatically. Perhaps we need better wording there. Can you try uploading a report please and we can have a look at what's gone wrong with the EFI loader.

**Pull request merged**